music_assistant/server/providers/tidal/__init__.py

```diff
@@ -0,0 +1,382 @@
+        tidal_session = TidalSession(config=tidal_config)
+        # load tokens from config and add tidal session to provider instance.
+        self._tidal_session = await validate_token_and_refresh(
+            tidal_session, self.mass, self.config
+        )
+        self._tidal_user_id = self.config.get_value(CONF_USER_ID)
+
+    @property
+    def supported_features(self) -> tuple[ProviderFeature, ...]:
+        """Return the features supported by this Provider."""
+        return (
+            ProviderFeature.LIBRARY_ARTISTS,
+            ProviderFeature.LIBRARY_ALBUMS,
+            ProviderFeature.LIBRARY_TRACKS,
+            ProviderFeature.LIBRARY_PLAYLISTS,
+            ProviderFeature.ARTIST_ALBUMS,
+            ProviderFeature.ARTIST_TOPTRACKS,
+            ProviderFeature.SEARCH,
+            ProviderFeature.LIBRARY_ARTISTS_EDIT,
+            ProviderFeature.LIBRARY_ALBUMS_EDIT,
+            ProviderFeature.LIBRARY_TRACKS_EDIT,
+            ProviderFeature.LIBRARY_PLAYLISTS_EDIT,
+            ProviderFeature.PLAYLIST_CREATE,
+            ProviderFeature.SIMILAR_TRACKS,
+            ProviderFeature.BROWSE,
+            ProviderFeature.PLAYLIST_TRACKS_EDIT,
+        )
+
+    async def search(
+        self, search_query: str, media_types=list[MediaType] | None, limit: int = 5
+    ) -> SearchResults:
+        """Perform search on musicprovider.
+
+        :param search_query: Search query.
+        :param media_types: A list of media_types to include. All types if None.
+        :param limit: Number of items to return in the search (per type).
+        """
+        search_query = search_query.replace("'", "")
+        results = await search(self._tidal_session, search_query, media_types, limit)
+        parsed_results = SearchResults()
+        if results["artists"]:
+            for artist in results["artists"]:
+                parsed_results.artists.append(await parse_artist(artist))
+        if results["albums"]:
+            for album in results["albums"]:
+                parsed_results.albums.append(await parse_album(album))
+        if results["playlists"]:
+            for playlist in results["playlists"]:
+                parsed_results.playlists.append(await parse_playlist(playlist))
+        if results["tracks"]:
+            for track in results["tracks"]:
+                parsed_results.tracks.append(await parse_track(track))
+        return parsed_results
+
+    async def get_library_artists(self) -> AsyncGenerator[Artist, None]:
+        """Retrieve all library artists from Tidal."""
+        artists_obj = await get_library_artists(self._tidal_session, self._tidal_user_id)
+        for artist in artists_obj:
+            yield parse_artist(tidal_provider=self, artist_obj=artist)
+
+    async def get_library_albums(self) -> AsyncGenerator[Album, None]:
+        """Retrieve all library albums from Tidal."""
+        albums_obj = await get_library_albums(self._tidal_session, self._tidal_user_id)
```

Comment:
replace the "self._tidal_session" with your helper method that checks the token and returns the session